lib/services/local/inventory_test.go

```diff
@@ -0,0 +1,245 @@
+	require.NoError(t, err)
+
+	raw1, err := presence.CompareAndSwapInstance(ctx, instance1, nil)
+	require.NoError(t, err)
+
+	// verify that "create" style compare and swaps are now rejected
+	_, err = presence.CompareAndSwapInstance(ctx, instance1, nil)
+	require.Error(t, err)
+	require.True(t, trace.IsCompareFailed(err))
+
+	// get the inserted instance
+	instances, err := stream.Collect(presence.GetInstances(ctx, types.InstanceFilter{}))
+	require.NoError(t, err)
+	require.Len(t, instances, 1)
+
+	// verify that expiry and last_seen are automatically set to expected values.
+	exp1 := instances[0].Expiry()
+	seen1 := instances[0].GetLastSeen()
+	require.False(t, exp1.IsZero())
+	require.False(t, seen1.IsZero())
+	require.Equal(t, seen1, presence.Clock().Now().UTC())
+	require.Equal(t, exp1, seen1.Add(apidefaults.ServerAnnounceTTL))
+
+	require.True(t, exp1.After(presence.Clock().Now()))
+	require.False(t, exp1.After(presence.Clock().Now().Add(apidefaults.ServerAnnounceTTL*2)))
+
+	// update the instance control log
+	instance1.AppendControlLog(types.InstanceControlLogEntry{
+		Type: "testing",
+		ID:   1,
+		TTL:  time.Hour * 24,
+	})
+	instance1.SyncLogAndResourceExpiry(apidefaults.ServerAnnounceTTL)
+
+	// verify expected increase in ttl to accommodate custom log entry TTL (sanity check
+	// to differentiate bugs in SyncLogAndResourceExpiry from bugs in presence/backend).
+	require.Equal(t, seen1.Add(time.Hour*24), instance1.Expiry())
+
+	// perform normal compare and swap using raw value from previous successful call
+	_, err = presence.CompareAndSwapInstance(ctx, instance1, raw1)
+	require.NoError(t, err)
+
+	// verify that raw value from previous successful CaS no longer works
+	_, err = presence.CompareAndSwapInstance(ctx, instance1, raw1)
+	require.Error(t, err)
+	require.True(t, trace.IsCompareFailed(err))
+
+	// load new instance state
+	instances2, err := stream.Collect(presence.GetInstances(ctx, types.InstanceFilter{}))
+	require.NoError(t, err)
+	require.Len(t, instances2, 1)
+
+	// ensure that ttl and log were preserved
+	require.Equal(t, seen1.Add(time.Hour*24), instances2[0].Expiry())
+	require.Len(t, instances2[0].GetControlLog(), 1)
+}
+
+// TestInstanceFiltering tests basic filtering options. A sufficiently large
+// instance count is used to ensure that queries span many pages.
+func TestInstanceFiltering(t *testing.T) {
+	const count = 100_000
+	t.Parallel()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// NOTE: backend must be memory, since parallel subtests are used (makes correct cleanup of
+	// filesystem state tricky).
+	backend, err := memory.New(memory.Config{
+		Context: ctx,
+		Clock:   clockwork.NewFakeClock(),
+	})
+	require.NoError(t, err)
+
+	defer backend.Close()
+
+	presence := NewPresenceService(backend)
+
+	// store an odd and an even uuid for later use in queries
+	var evenID, oddID string
+
+	evenServices := []types.SystemRole{"even"}
```

Comment:
We're allowing any arbitrary value for system role now?

Reply:
Teleport only generates certs with known values, but the inventory system allows heartbeat of arbitrary values if those values are on your cert.

This is necessary in order to ensure that if you have to downgrade from a newer version where a new service name was added, you don't have your whole instance bricked because there is an unknown service name on the cert.